mbed_devices/_internal/resolve_target.py

```diff
@@ -84,5 +99,9 @@ def _get_all_htm_files_contents(directories: Iterable[pathlib.Path]) -> List[str
 
 
 def _is_htm_file(file: pathlib.Path) -> bool:
+    """Checks whether the file looks like an Mbed HTM file.
+
+    Note that if the file is manually manipulated the file may sometimes appears to be present but not be accessible.
+    """
     extensions = [".htm", ".HTM"]
-    return file.suffix in extensions and not file.name.startswith(".")
+    return file.suffix in extensions and not file.name.startswith(".") and os.access(file, os.R_OK)
```

Comment:
Could you add a test for this? How does `os.access(file, os.R_OK)` present itself? Could we not rely on exception instead?

Rationale: untested bugfix will hurt us in the future, code coverage lies.

Reply:
I'd prefer we just caught an `OSError` here (if possible) instead of using `R_OK` as that's rather opaque.

Reply:
@muchzill4 I can add a test, it seems to manifest because of a problem with the file system. The directory is listed but not actually present. It may also occur if the file cannot be read due to permission, but I haven't tried this.

@rwalton-arm The exception occurs inside the list comprehension so there are two other options:
- Unpacking the creation of the list and catch the exception - the code is not so nice doing this
- Catch the error at a higher level and abort the read for all files is there is a read error - less robust

Reply:
How about:
```
try:
   return file.suffix in extensions and not file.name.startswith('.')
except SOME_ERROR:
   return False
```

I think that was the suggestion?

Reply:
That's what I was thinking of. I'm not a fan of using `os.access` as it can potentially give misleading results. The docs also mention the potential to create security holes as it's messing with UID/GID to verify access rights. Essentially it's a can of very complicated permissions related worms. Even the python docs recommend using EAFP techniques as suggested. I'd suspect there's also the possibly we would encounter subtle differences in behaviour on Windows, as it has an even more complicated permissions model.

Reply:
@muchzill4 Unfortunately the exception doesn't occur there, it occurs in the list comprehension calling this function when the read is performed.

Reply:
Ah, I see now. I got confused by where the `os.access` check was put. For some reason I thought it happens when file is interrogated for its name parts, which makes _absolutely no sense_. 🤦‍♂ 

With that in mind, I think this fix should live closer to the origin of the problem (`file.read_text()` is where it happens, right?).

I think handling file readability does not belong in `_is_htm_file` but should instead be shifted to where comprehension happens. It'll probably no longer be a comprehension, but a `for` loop. We could also have an appropriate log message to make the debug logs even better!

What I'm thinking:

```
all_files = ...
all_htm_files = (file for file in all_files if _is_htm_file(file)
htm_files_contents = []
for htm_file in htm_files:
   try:
      html_files_contents.append(htm_file.read_text())
   except SOME_ERROR:
      logging.approriate_level(f"Can't read {htm_file}.")
return htm_files_contents
```

Reply:
Indeed, this is my option:

- Unpacking the creation of the list and catch the exception - the code is not so nice doing this

Reply:
Ah, that makes sense! I can only speak code snippets @madchutney . 😎

Reply:
@muchzill4 I have done something similar to your suggestion but to facilitate testing I had to create a separate function. Not too bad although a lot more code.

Reply:
Thanks. I agree it's more verbose. I think it's worth paying that price for better coverage.